exercises/src/main/scala/fpinscala/parallelism/Nonblocking.scala

```diff
@@ -123,42 +139,64 @@ object Nonblocking {
      */
     def choice[A](p: Par[Boolean])(t: Par[A], f: Par[A]): Par[A] =
       es => new Future[A] {
-        def apply(cb: A => Unit): Unit =
+        def apply(cb: A => Unit)(kErr: Throwable => Unit): Unit =
           p(es) { b =>
-            if (b) eval(es) { t(es)(cb) }
-            else eval(es) { f(es)(cb) }
-          }
+            if (b) eval(es) { t(es)(cb)(kErr) }
+            else eval(es) { f(es)(cb)(kErr) }
+          } { err => throw err }
       }
 
-    def choiceN[A](p: Par[Int])(ps: List[Par[A]]): Par[A] = ???
-
-    def choiceViaChoiceN[A](a: Par[Boolean])(ifTrue: Par[A], ifFalse: Par[A]): Par[A] =
-      ???
+    def choiceN[A](p: Par[Int])(ps: List[Par[A]]): Par[A] =
+      es => new Future[A] {
+        def apply(k: A => Unit)(kErr: Throwable => Unit): Unit =
+          p(es)(ps(_)(es)(k)(kErr))(throw _)
+      }
 
-    def choiceMap[K,V](p: Par[K])(ps: Map[K,Par[V]]): Par[V] =
-      ???
+    def choiceViaChoiceN
+      [A]
+      (a: Par[Boolean])
+      (ifTrue: Par[A], ifFalse: Par[A]):
+      Par[A] =
+      choiceN(map(a)(if (_) 0 else 1))(List(ifTrue, ifFalse))
```

Comment:
@gpoirier unfortunately some parts of Scala syntax are difficult to break across lines if we want to keep inside 72 char columns.